### PR TITLE
Feature: Add edit tags button to the preview pane

### LINKED ISSUE #12821: Feature: Tag section on Detials Pane, should include "Add Tags" button
### What feature or improvement do you think would benefit Files?

As mentioned in #12810 when a file or folder has no tags, a tags section still shows in the detail pane.

image

Rather than hiding this when no tags have been attached to the item, we could put an "Add Tag" or "Add Tags" button.

image
image

### Requirements

- This will simplify adding tags to items, and emphasise the feature to people who are not familiar with it.

### Files Version

2.5.10

### Windows Version

22621.1928

### Comments

_No response_

### LINKED ISSUE #12947: Bug: Tags are not displayed in the Preview Pane if no item is selected
### Description

If no item is selected, the Preview Pane will show details of the current folder. The Preview Pane won't show its tags (even if it has some)

### Steps To Reproduce

1. Select a tagged folder
2. Open the preview pane
3. See that it displays selected folder's tags
4. Open that folder
5. See that the tags section won't be rendered (no tags are detected)

### Requirements

Tags (if any) should always be displayed in the preview pane

### Files Version

Files Preview 2.5.14.0

### Windows Version

Windows 11 Pro 22H2 22621.1992

### Log File

Selected item:
Screenshot 2023-07-14 135904

No item selected:
Screenshot 2023-07-14 135854

### PR DESCRIPTION
**Pull request by ferrariofilippo**
**Resolved / Related Issues**
- [x] Were these changes approved in an issue or discussion with the project maintainers? In order to prevent extra work, feature requests and changes to the codebase must be approved before the pull request will be reviewed. This prevents extra work for the contributors and maintainers.
   Closes #12821 
   Closes #12947 

**Validation**
How did you test these changes?
- [x] Did you build the app and test your changes?
- [ ] Did you check for accessibility? You can use Accessibility Insights for this.
- [ ] Did you remove any strings from the en-us resource file?
   - [ ] Did you search the solution to see if the string is still being used? 
- [x] Did you implement any design changes to an existing feature?
   - [x] Was this change approved?
- [x] Are there any other steps that were used to validate these changes?
   1. Open app and select a folder
   2. Open the preview pane
   3. Play with tags

**Screenshots (optional)**
Screenshot 2023-07-13 234208

### CODE REVIEW

**Comment by yaira2** [comment]
We may want to consider hiding this on items without tags, this will be consistent with other properties that are hidden when empty.

**Comment by mdtauk** [comment]
I am assuming this is using the same flyout as is used elsewhere in the app - so there is no code duplication...

**Comment by ferrariofilippo** [comment]
> I am assuming this is using the same flyout as is used elsewhere in the app - so there is no code duplication...

Yes

**Comment by ferrariofilippo** [comment]
> We may want to consider hiding this on items without tags, this will be consistent with other properties that are hidden when empty.

I made it always visible since that was the required behaviour in the issue
> Rather than hiding this when no tags have been attached to the item, we could put an "Add Tag" or "Add Tags" button.

Anyway, I think we should keep it even if there are no tags, it's quicker to edit tags using that button than using the context menu.
Still we could hide it if the user has turned off `Show tags in context menu`

**Comment by ferrariofilippo** [comment]
image

**Comment by yaira2** [comment]
@ferrariofilippo Can you display the tags horizontally (and wrap onto the next line if there isn't enough room)?

**Comment by mdtauk** [comment]
> @ferrariofilippo Can you display the tags horizontally (and wrap onto the next line if there isn't enough room)?

Two per row I think will be enough, if they take up 50% of the space each, This way the tag icons below will align to the ones above, and it wont look staggered.

**Comment by yaira2** [comment]
> Still we could hide it if the user has turned off

It's okay, we can always show it.

**Comment by ferrariofilippo** [comment]
@yaira2 do you know any control better than `ItemsWrapGrid`? Using that you have to set manually items width

**Comment by yaira2** [comment]
@ferrariofilippo We can use a GridView with a custom template, the Community Toolkit has a couple of templates with the desired behavior.

**Comment by ferrariofilippo** [comment]
> @ferrariofilippo We can use a GridView with a custom template, the Community Toolkit has a couple of templates with the desired behavior.

Did you mean this?
https://learn.microsoft.com/en-us/windows/apps/design/controls/item-templates-gridview#icon-and-text

**Comment by yaira2** [comment]
@ferrariofilippo https://learn.microsoft.com/en-us/dotnet/api/microsoft.toolkit.uwp.ui.controls.wraplayout?view=win-comm-toolkit-dotnet-7.0
image

**Comment by ferrariofilippo** [comment]
> @ferrariofilippo https://learn.microsoft.com/en-us/dotnet/api/microsoft.toolkit.uwp.ui.controls.wraplayout?view=win-comm-toolkit-dotnet-7.0

Using that you can't display two tags/line, is that fine?
image

**Comment by yaira2** [comment]
> Using that you can't display two tags/line, is that fine?

What do you mean? The screenshot looks like the desired behavior...

**Comment by ferrariofilippo** [comment]
> Two per row I think will be enough, if they take up 50% of the space each, This way the tag icons below will align to the ones above, and it wont look staggered.

I thought we were going for this

**Review by yaira2 — APPROVED** [review]
LGTM

File: src/Files.App/UserControls/Pane/PreviewPane.xaml [hunk 1]
```diff
@@ -269,6 +271,63 @@
 					</ItemsControl.ItemTemplate>
 				</ItemsControl>
 
+				<!--  Tags List  -->
+				<StackPanel
+					x:Name="DetailsTagsList"
+					Spacing="4"
```
Comment (yaira2) [inline src/Files.App/UserControls/Pane/PreviewPane.xaml:277]: Can x:Load be false when there aren't any tags? Users will still be able to add tags from the menu flyout.

**Comment by yaira2** [comment]
image
I think we can improve the design by removing the "edit tag" button and instead, adding a "plus" icon to the GridView.

**Comment by ferrariofilippo** [comment]
> I think we can improve the design by removing the "edit tag" button and instead, adding a "plus" icon to the GridView.

Where should it be? Before/after the tags? Next to the header?

**Comment by yaira2** [comment]
Ideally it'll be in the list of tags, is this possible?

**Comment by ferrariofilippo** [comment]
I'm not sure we can do that but I'll see

**Comment by ferrariofilippo** [comment]
@yaira2 I think we can do that using something like this, is it fine?
```
public class TagsListItem
{
	public object Item { get; set; }

	public bool IsButton => object is Button;

	public bool IsTag => object is Tag;
}

<DataTemplate>
	<StackPanel>
		<TagLayout x:Load="{x:Bind obj.IsTag, Mode=OneWay}" other_props />
		<Button x:Load="{x:Bind obj.IsButton, Mode=OneWay}" />
	</StackPanel>
</DataTemplate>

```

**Comment by yaira2** [comment]
Seems like a bit of a hack but it's simple enough so we might as well.

**Comment by yaira2** [comment]
@ferrariofilippo even better, what if we added a "plus" icon next to the "Tags" header?

**Comment by ferrariofilippo** [comment]
Does it look better using `OpacityIcon` or using the `Glyph`?

Screenshot 2023-07-19 102839
Screenshot 2023-07-19 103318

**Comment by Lukiluc29** [comment]
hello, sorry to embed myself in your discussion but it would not be better to put the button after the tags instead of putting it next to "tags"
254524240-81867ae9-fe5e-4dd7-ac09-e74a728721b2 (1)

**Comment by ferrariofilippo** [comment]
> Hello, sorry to embed myself in your discussion but it would not be better to put the button after the tags instead of putting it next to "tags" 254524240-81867ae9-fe5e-4dd7-ac09-e74a728721b2 (1)

This was the original idea but it's quite a mess to implement, we would need some hacky stuff

**Comment by Lukiluc29** [comment]
It's a shame because the original idea is much more beautiful than the one you have just proposed, as, for me, it would look out of place next to the 'tags'.

**Comment by ferrariofilippo** [comment]
@yaira2 what do you think? Both the solutions work properly. I can push the one you prefer

**Comment by yaira2** [comment]
> Is it going to be an edit button however, or an add tag button.

For now it's going to be the edit tag button.

**Comment by mdtauk** [comment]
image
What about something like this.  You can quick remove tags, or press the button to do more and add new ones.

**Comment by Lukiluc29** [comment]
> image What about something like this. You can quick remove tags, or press the button to do more and add new ones.

It's a great concept that will be beautiful once implemented.

**Pull request merged**